Added kirby-chart-wip component; allowing for bar & column charts

MadsBuchmann opened the pull request:
## Which issue does this PR close?

This PR: 
* closes #1572 
* closes #1574 
* closes #1652 

## What is the new behavior?
A `kirby-chart-wip` component has been added to Kirby Design System which uses chart.js and not highcharts which has been used until now. 

Instead of explaining it all here - please spin up the development server on your machine and see the _Chart Wip_ cookbook page, which explains this component. 

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

<!-- If this PR contains a breaking change, replace this paragraph with a description of the impact and migration path for existing applications  -->

## Are there any additional context?
### BE AWARE: This PR should not be merged to master yet 
After the PR has passed code review #1652  should be carried out, wherein the old Highcharts charts is deprecated



<!-- Replace this paragraph with any additional context e.g, explanations, links or screenshots (if any) -->

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](../CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [X] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [X] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [X] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [X] Request that the changes are code-reviewed 
- [x] Request that the changes are [UX reviewed](./CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be automatically merged to master via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/chart-wip.types.ts:
```diff
@@ -0,0 +1,13 @@
+export type ChartDataset = ChartJSDataset<'bar'>;
+export type ChartData = ChartDataset[] | number[];
+
+export function isNumberArray(value: any): value is number[] {
```
Is there a better place to store type helpers like this one?

RasmusKjeldgaard replied:
As long as it is only used by the chart, I think this is an ok place. We can move it to /helpers/ if we ever reuse it in another context.

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/index.ts:
```diff
@@ -0,0 +1,7 @@
+  ChartData as ChartWipData,
+  ChartDataset as ChartWipDataset,
+} from './chart-wip.types';
+export { ChartOptions as ChartWipOptions } from 'chart.js';
```
Once highcharts is removed the 'x as y' shall be removed.

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/chart-wip-example/examples/chart-wip-example-column-stacked.component.ts:
```diff
@@ -0,0 +1,67 @@
+import { Component } from '@angular/core';
+import { ChartOptions } from 'chart.js';
```
Should we ideally expose this through Kirby? 

We ended up doing a 'proxy' of the date-fns API for some properties in the Calendar component when it was reimplemented recently, and I like that approach as it shows what properties we _intend_ consumers to modify. And we can still reference the docs for the chart.js API directly, as it is 1 to 1.

So, something similar to [line 44](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/calendar/calendar.component.ts#L44) and [line 29](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/calendar/calendar.component.ts#L29), where we export a the Locale from date-fns so it can be referenced through Kirby.

MadsBuchmann replied:
That would be much prettier - I agree! 
Will get that done.

RasmusKjeldgaard reviewed (commented):
Dude... Awesome work 🎉  🎉 

Love how simple the cookbook markup examples are, because of the sensible defaults!  

To keep the looks of the new chart component as close to the original as possible, we should consider the following: 
- Implement this as a [mixed chart](https://www.chartjs.org/docs/latest/charts/mixed.html), so an additional data-set, or single value, can be applied to create the dotted line of the existing bar chart. So essentially extend the api of this specific component with the ability to also draw a line.  
- **Or** expose the line chart, and create a cookbook example showing how to create a mixed chart type with a dotted line. 

We also need to look at the [bar thickness or max bar thickness](https://www.chartjs.org/docs/latest/charts/bar.html#barthickness) so it is in line with the existing solution

MadsBuchmann commented on libs/designsystem/package.json:
```diff
@@ -2,6 +2,7 @@
   "dependencies": {
+    "chart.js": "3.3.2",
     "@ionic/angular": "5.5.2",
     "highcharts": "8.0.4",
```
Yup! Should be carried out as part of #1413 :D

MadsBuchmann commented on apps/cookbook/src/app/showcase/chart-showcase/chart-showcase.component.html:
```diff
@@ -1,7 +1,6 @@
-    <cookbook-chart-example-donut #donut> </cookbook-chart-example-donut>
-  </cookbook-example-viewer>
+  <cookbook-chart-example-donut #donut> </cookbook-chart-example-donut>
+  <cookbook-example-viewer [html]="donut.template"> </cookbook-example-viewer>
```
Yup - it has been reverted 🦆

MadsBuchmann commented:
> Dude... Awesome work 🎉 🎉
> 
> Love how simple the cookbook markup examples are, because of the sensible defaults!
> 
> To keep the looks of the new chart component as close to the original as possible, we should consider the following:
> 
> * Implement this as a [mixed chart](https://www.chartjs.org/docs/latest/charts/mixed.html), so an additional data-set, or single value, can be applied to create the dotted line of the existing bar chart. So essentially extend the api of this specific component with the ability to also draw a line.
> * **Or** expose the line chart, and create a cookbook example showing how to create a mixed chart type with a dotted line.
> 
> We also need to look at the [bar thickness or max bar thickness](https://www.chartjs.org/docs/latest/charts/bar.html#barthickness) so it is in line with the existing solution

I have added an example to the cookbook page showing how to draw a line in a bar chart for good measure as the ability to mix charts is pretty cool. 

However I think what you're requesting is a plotline and not a separate type. I suppose we should integrate [annotations](https://www.chartjs.org/chartjs-plugin-annotation/guide/) and perhaps extend the api with the following two input properties: 
- `plotLines` here you can pass sets of coordinates where you want plotlines to be drawn e.g. ```plotlines = [ [[3,1], [3,2]], [[6,5], [6,3]]``` (might not actually go for a 3d array as that introduces a lot of parenthesis - but it's 17:03 and I wanna go home. Will think deeper about it tomorrow :) ).  
- `customAnnotations`:  Much like `customOptions` this should allow the consumer to go ham using the annotations plugins.

`plotLines` is the easy way to annotate your charts where using `customAnnotations` is a bit more complex but gives you a higher degree of freedom.

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/utils.ts:
```diff
@@ -0,0 +1,16 @@
+  return deepMergeAll([{}, ...objectsWithoutUndefined], deepMergeOptions);
+}
+
+export function deepCopy(obj: any) {
+  return JSON.parse(JSON.stringify(obj));
+}
```
Is there a more appropriate place to store this function? It is not necessarily only useful in the context of this component.

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/utils.ts:
```diff
@@ -0,0 +1,16 @@
+import { all as deepMergeAll, Options as deepMergeOptions } from 'deepmerge';
+
+export function deepMergeObjects(...objects: any[]) {
+  const overwriteMerge = (_target: any[], source: any[], _options: deepMergeOptions) => source;
+  const deepMergeOptions: deepMergeOptions = {
+    arrayMerge: overwriteMerge,
+  };
+
+  const objectsWithoutUndefined = objects.filter((object) => object !== undefined);
+
+  return deepMergeAll([{}, ...objectsWithoutUndefined], deepMergeOptions);
+}
```
Is there a more appropriate place to store this function? It is not necessarily only useful in the context of this component.

RasmusKjeldgaard commented on libs/designsystem/src/lib/components/chart-wip/chart-wip.component.spec.ts:
```diff
@@ -0,0 +1,146 @@
+
+          spectator.setInput(property as any, newValue);
+
+          expect(updateFnSpy).toHaveBeenCalledTimes(1);
```
This seems like it checks that the related function has been called once, but not that any of the other functions have not been called.

If that is the intention of the test, I think it should be refactored to reflect that. E.g. have two scenarios `old` and `new` where the component values `component.type` etc. can be compared before and after `setInput`.

Otherwise, I think we should remove 'only' in `it('should only update ${property} ...`

RasmusKjeldgaard replied:
Or create multiple spies like line 107-112, and expect all other than the current one to have been called 0 times.

MadsBuchmann replied:
I've removed `only` from the description &  created a separate tests that checks that none of the other properties are set

RasmusKjeldgaard commented on libs/designsystem/src/lib/components/chart-wip/chart-wip.component.spec.ts:
```diff
@@ -0,0 +1,146 @@
+    });
+  });
+
+  describe('when "type" is not provided', () => {
+    it('should default to correct type', () => {
+      expect(component.type).toBe('column');
+    });
+  });
```
Could we move this closer to the top? I think it fits well with `should have correct default height`-tests and the ones like it.

MadsBuchmann replied:
Has been done :D

RasmusKjeldgaard commented on libs/designsystem/src/lib/components/chart-wip/chart-js/chart-js.service.ts:
```diff
@@ -0,0 +1,232 @@
+    to make it optional for consumer */
+    const labels = !dataLabels ? this.createBlankLabels(datasets) : dataLabels;
+    const typeConfig = this.getTypeConfig(type);
+    //TODO: do we need deep merge here?
```
Remember this todo. 

My take is that it can't hurt (famous last words), and then it would be nicely aligned with `createOptionsObject()`

MadsBuchmann replied:
I opted for deep merge :D

RasmusKjeldgaard commented on libs/designsystem/src/lib/components/chart-wip/chart-js/chart-js.service.spec.ts:
```diff
@@ -0,0 +1,865 @@
+        ];
+
+        chartJSService.renderChart({
+          targetElement: canvasElement,
+          type: 'bar',
+          data: [1, 2, 3],
+          dataLabels: ['one', 'two', 'three'],
```
This part of the object is reused quite a lot, consider extracting to a variable. 
```suggestion
          ...options,
```

MadsBuchmann replied:
I get your points - but in order to keep the tests [isolated & flat](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test#the-good-test-is-isolated-and-flat) I think the correct way about it is to duplicate the options (eventough it is very WET)

RasmusKjeldgaard replied:
Good point, I was probably still a bit in component code review mode when I wrote this 😄

RasmusKjeldgaard reviewed (changes requested):
My dude, big brain work. Looks very good all around 🎉 🏆 
I mostly have some questions and discussion points for some possible improvements.

MadsBuchmann commented on apps/cookbook/src/app/showcase/chart-wip-showcase/chart-wip-showcase.component.html:
```diff
@@ -0,0 +1,290 @@
+  href="https://www.chartjs.org/docs/master/configuration/responsive.html#configuration-options"
+  ><code>maintainAspectRatio</code></a
+>
+to <code>false</code>.
```
You're 100% right!

MadsBuchmann commented on apps/cookbook/src/app/showcase/chart-wip-showcase/chart-wip-showcase.component.html:
```diff
@@ -0,0 +1,290 @@
+</p>
+<p>
+  As a fallback for screen readers structured content can be slotted that a screen reader will read
+  aloud; see the below example.
```
Tanks!

MadsBuchmann commented on apps/cookbook/src/app/examples/chart-wip-example/examples/chart-wip-example-interaction.component.ts:
```diff
@@ -0,0 +1,68 @@
+import { Component } from '@angular/core';
+import { ActiveElement, Chart, ChartEvent } from 'chart.js';
+
+import { ChartOptions } from '@kirbydesign/designsystem';
```
Yup yup yup!

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/chart-js/configured-chart-js.ts:
```diff
@@ -0,0 +1,33 @@
+
+/* Chart.defaults is read only; set each key modified in 
+   CHART_GLOBAL_DEFAULTS manually */
+const mergedDefaults = deepMergeObjects(Chart.defaults, CHART_GLOBAL_DEFAULTS) as any;
```
I opted for removing the typecast - it does not do anything anyways 😄

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/utils.ts:
```diff
@@ -0,0 +1,16 @@
+import { all as deepMergeAll, Options as deepMergeOptions } from 'deepmerge';
+
+export function deepMergeObjects(...objects: any[]) {
```
Done!

MadsBuchmann commented on libs/designsystem/src/lib/components/chart-wip/chart-wip.component.spec.ts:
```diff
@@ -0,0 +1,146 @@
+    const element = spectator.query('.chart-container') as HTMLElement;
+    expect(element).not.toHaveComputedStyle({ height: customHeight });
+
+    element.style.setProperty('--kirby-chart-height', customHeight);
```
Good catch! - I've rewritten the test to use host elements instead.

RasmusKjeldgaard reviewed (approved):
Wohooo, looking good!

MadsBuchmann commented:
@RasmusKjeldgaard I've moved the functions from `utils.ts` to the `.../lib/helpers` folder instead as separate files. 
Also I've renamed the `deep-merge.ts` files to `merge-deep.ts` as it drove me insane that the function inside was called `mergeDeep` 🦈

RasmusKjeldgaard reviewed (dismissed):
Here we go, amazing job!

MadsBuchmann commented:
@RasmusKjeldgaard it needs to be approved by you again after version bump :D
